src/H5private.h

```diff
@@ -338,7 +347,10 @@
 /* Check if a read of size bytes starting at ptr would overflow past
  * the last valid byte, pointed to by buffer_end.
  */
-#define H5_IS_BUFFER_OVERFLOW(ptr, size, buffer_end) (((ptr) + (size)-1) > (buffer_end))
+#define H5_IS_BUFFER_OVERFLOW(ptr, size, buffer_end)                                                         \
+    (((ptr) > (buffer_end)) ||                             /* Bad precondition */                            \
+     ((ptrdiff_t)(size) > (((buffer_end) - (ptr)) + 1)) || /* Typical overflow */                            \
+     ((intptr_t)(size) < 0))                               /* Negative 'size' would wrap 'ptr' */
```

Comment:
So I think it's technically possible to have a valid buffer so large that ptrdiff_t wraps to negative when computing the size. This could be an issue on 32 bit systems that are pushing the limits of their architecture. So size could be negative when cast to ptrdiff_t be valid. I don't know if the rest of the library would handle this case correctly.

Reply:
Hmm yes I think this is a bit more in line with what I was thinking.

Reply:
After a little research though, it seems some compilers often remove a `ptr + len < ptr` type check since it's UB anyway.

Reply:
I suggest getting my notes for files that tripped overflows from Dana and making certain that any change to what I worked out last year are still protecting from the related corrupted files.  Some of them were pretty obscure... :-/

Reply:
Definitely plan to test against the CVE files. I think it's mostly a matter of needing to check `size` against the max of the signed type it's being cast into first, and then casting it and checking against < 0 before then doing the pointer subtraction and comparison. Unfortunately it's difficult to do this in a nice way since the macro has no type information and it's not always a `size_t` value that's being passed in.

Reply:
Rewrote this slightly. It's basically the same thing, but using `ptrdiff_t` instead of `intptr_t`, checking against `PTRDIFF_MAX` before casting and also doing the check for negative values before the typical overflow check. Passes all current CVE tests, including the ones specifically around the changes in this PR.